flink-connector-mongodb/src/main/java/org/apache/flink/connector/mongodb/table/MongoDynamicTableSink.java

```diff
@@ -0,0 +1,124 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.mongodb.table;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.connector.mongodb.common.config.MongoConnectionOptions;
+import org.apache.flink.connector.mongodb.sink.MongoSink;
+import org.apache.flink.connector.mongodb.sink.config.MongoWriteOptions;
+import org.apache.flink.connector.mongodb.table.converter.RowDataToBsonConverters;
+import org.apache.flink.connector.mongodb.table.converter.RowDataToBsonConverters.RowDataToBsonConverter;
+import org.apache.flink.connector.mongodb.table.serialization.MongoRowDataSerializationSchema;
+import org.apache.flink.table.connector.ChangelogMode;
+import org.apache.flink.table.connector.sink.DynamicTableSink;
+import org.apache.flink.table.connector.sink.SinkV2Provider;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.util.function.SerializableFunction;
+
+import org.bson.BsonValue;
+
+import javax.annotation.Nullable;
+
+import java.util.Objects;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** A {@link DynamicTableSink} for MongoDB. */
+@Internal
+public class MongoDynamicTableSink implements DynamicTableSink {
+
+    private final MongoConnectionOptions connectionOptions;
+    private final MongoWriteOptions writeOptions;
+    @Nullable private final Integer parallelism;
+    private final DataType physicalRowDataType;
+    private final SerializableFunction<RowData, BsonValue> keyExtractor;
+
+    public MongoDynamicTableSink(
+            MongoConnectionOptions connectionOptions,
+            MongoWriteOptions writeOptions,
+            @Nullable Integer parallelism,
+            DataType physicalRowDataType,
+            SerializableFunction<RowData, BsonValue> keyExtractor) {
+        this.connectionOptions = checkNotNull(connectionOptions);
+        this.writeOptions = checkNotNull(writeOptions);
+        this.parallelism = parallelism;
+        this.physicalRowDataType = checkNotNull(physicalRowDataType);
+        this.keyExtractor = checkNotNull(keyExtractor);
+    }
+
+    @Override
+    public ChangelogMode getChangelogMode(ChangelogMode requestedMode) {
+        return ChangelogMode.upsert();
```

Comment:
Shouldn't the changelog mode depend on the primary key? If we are not in sync with `AppendOnlyKeyExtractor` an update-before and update-after might be partitioned as separate entities. When rereading they would appear as two independent records and thus source and sink do not consume/produce the same data.

Reply:
Yes, changelog mode depend on the primary key.
If user explicitly declares a primary key in flink sql, we use `MongoKeyExtractor` to extract them as `_id` of MongoDB. If not, we use `AppendOnlyKeyExtractor` to pass MongoDB a null `_id` and let MongoDB automatically generate ObjectId.

Considering that there will be some situations where the primary key definition is forgotten, we may need to provide some configuration to explicitly specify whether to write as `append-only` or `upsert` mode so that we can do some verification at startup.

Reply:
What I had in mind was something like:
```
if (hasPrimaryKey(schema)) {
  return ChangelogMode.upsert();
} else {
  return ChangelogMode.insertOnly();
}
```

This would ensure that the planner will not allow updates to be written into a sink without primary key defined.